File: sdks/csharp/PgChangeFeed.Client/Sse/SseFrameParser.cs
namespace PgChangeFeed.Client.Sse;

/// <summary>
/// One complete SSE frame — the <c>event:</c> name and the <c>data:</c>
/// payload accumulated up to the blank line that terminates it.
/// </summary>
public sealed record SseFrame(string Name, string Data);

/// <summary>
/// Splits Server-Sent-Events frames from a line source into
/// <see cref="SseFrame"/> values — plain line processing over the runtime's
/// standard library, no further dependency.
///
/// A frame ends with the blank line the server writes after the <c>event:</c>
/// and the <c>data:</c> line; it separates two consecutive events. When the
/// source is exhausted before a frame is complete
/// (<see cref="TextReader.ReadLineAsync(System.Threading.CancellationToken)"/>
/// returns <c>null</c>), <see cref="ReadFrameAsync"/> also returns
/// <c>null</c>, and a frame already begun is discarded — an incomplete frame
/// is not an event.
/// </summary>
public static class SseFrameParser
{
    private const string EventPrefix = "event: ";
    private const string DataPrefix = "data: ";

    /// <summary>
    /// Reads the next complete frame, or returns <c>null</c> when the source
    /// ends before a frame is complete.
    /// </summary>
    public static async Task<SseFrame?> ReadFrameAsync(
        TextReader reader, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(reader);

        string? name = null;
        string? data = null;
        while (true)
        {
            var line = await reader.ReadLineAsync(cancellationToken).ConfigureAwait(false);
            if (line is null)
            {
                return null;
            }
            if (line.Length == 0)
            {
                if (name is null && data is null)
                {
                    continue;
                }
                return new SseFrame(name ?? "", data ?? "");
            }
            if (line.StartsWith(EventPrefix, StringComparison.Ordinal))
            {
                name = line[EventPrefix.Length..];
            }
            else if (line.StartsWith(DataPrefix, StringComparison.Ordinal))
            {
                data = line[DataPrefix.Length..];
            }
        }
    }
}

File: sdks/kotlin/pgchangefeed-kotlin/src/main/kotlin/io/github/pt9912/pgchangefeed/sse/SseFrameParser.kt
package io.github.pt9912.pgchangefeed.sse

/**
 * One complete SSE frame of the change stream — the `event:` name and the
 * `data:` payload accumulated up to the blank line that terminates it.
 */
internal data class SseFrame(val name: String, val data: String)

/**
 * Splits Server-Sent-Events frames from a line source into [SseFrame] values —
 * plain line processing over the JDK standard library, no further dependency.
 *
 * A frame ends with the blank line the server writes after the `event:` and
 * the `data:` line; it separates two consecutive events. When the source is
 * exhausted before a frame is complete ([next] returns `null`), [readFrame]
 * also returns `null`, and a frame already begun is discarded — an incomplete
 * frame is not an event.
 */
internal object SseFrameParser {
    private const val EVENT_PREFIX = "event: "
    private const val DATA_PREFIX = "data: "

    fun readFrame(next: () -> String?): SseFrame? {
        var name: String? = null
        var data: String? = null
        while (true) {
            val line = next() ?: return null
            if (line.isEmpty()) {
                if (name == null && data == null) {
                    continue
                }
                return SseFrame(name ?: "", data ?: "")
            }
            when {
                line.startsWith(EVENT_PREFIX) -> name = line.removePrefix(EVENT_PREFIX)
                line.startsWith(DATA_PREFIX) -> data = line.removePrefix(DATA_PREFIX)
            }
        }
    }
}
